org.eclipse.xtext/src/org/eclipse/xtext/nodemodel/impl/NodeModelBuilder.java

```diff
@@ -32,54 +30,58 @@
  */
 public class NodeModelBuilder {
 
-	private static class ArrayInterner<T> implements Interner<T[]> {
+	private static class ArrayInterner {
 
-		private static class ArrayAsList<T> extends AbstractList<T> implements RandomAccess {
-			final T[] array;
-			private int hashCode = -1;
+		private static class InternKey {
+			final EObject myGrammarElement;
+			final Object childGrammarElement;
+			
+			int hashCode = -1;
 
-			ArrayAsList(/* @Nullable */ T[] array) {
-				this.array = array;
+			InternKey(EObject myGrammarElement, Object childGrammarElement) {
+				this.myGrammarElement = myGrammarElement;
+				this.childGrammarElement = childGrammarElement;
 			}
-
-			@Override public int size() {
-				return array.length;
-			}
-
-			@Override /* @Nullable */ public T get(int index) {
-				return array[index];
+			
+			EObject[] joinElements() {
+				if(childGrammarElement instanceof EObject) {
+					return new EObject[] {myGrammarElement, (EObject) childGrammarElement};
+				} else {
+					return ObjectArrays.concat(myGrammarElement, (EObject[]) childGrammarElement);
+				}
 			}
 
-			@Override public int hashCode() {
+			@Override
+			public int hashCode() {
 				if (hashCode == -1)
-					hashCode = Arrays.hashCode(array);
+					hashCode = Objects.hash(myGrammarElement, childGrammarElement);
```

Comment:
Here we allocate an unnecessary array.

Reply:
The performance of hashCode really isn't an issue, and I'm pretty sure that hotspot has no problem folding a statically sized array, even in C1. I'm not boxing primitives either so.. I don't think this is worth it.

Reply:
What I learn from @cdietrich and his project: never underestimate the cost of hashcode computation ;) 
I'm fine with keeping this as is, as long as it does not show up in any relevant measurements.